apps/src/templates/progress/googlePlatformApiRedux.js

```diff
@@ -0,0 +1,139 @@
+import firehoseClient from '@cdo/apps/lib/util/firehose';
+
+/*
+ * The Google Classroom Share Button is only available through the google
+ * platform api, so we have to add it to our page on load. We manage loading
+ * in redux since multiple components need to observe the loading state.
+ */
+
+const START_LOADING_GAPI = 'googlePlatformApi/START_LOADING_GAPI';
+const FINISH_LOADING_GAPI = 'googlePlatformApi/FINISH_LOADING_GAPI';
+
+const startLoadingGapi = startTime => ({type: START_LOADING_GAPI, startTime});
+const finishLoadingGapi = success => ({type: FINISH_LOADING_GAPI, success});
+
+export const GOOGLE_PLATFORM_API_ID = 'GooglePlatformApiId';
+const LOAD_TIMEOUT_MILLIS = 15000;
+
+const initialState = {
+  loading: false,
+  loaded: false,
+  loadStartTime: null
+};
+
+export function loadGooglePlatformApi() {
+  return (dispatch, getState) => {
+    if (gapiReady()) {
+      dispatch(finishLoadingGapi(true));
+      return Promise.resolve();
+    } else {
+      dispatch(startLoadingGapi(Date.now()));
+      return new Promise((resolve, reject) => {
+        const promise = {resolve: resolve, reject: reject};
+        loadApi(dispatch, getState, promise);
+      });
+    }
+  };
+}
+
+function onLoadFinished(success, promise) {
+  return (dispatch, getState) => {
+    dispatch(finishLoadingGapi(success));
+    if (success) {
+      promise.resolve();
+    } else {
+      promise.reject('Google Platform API failed to load.');
+    }
+
+    const data = {
+      success: success,
+      load_time: elapsedLoadTimeSeconds(getState)
+    };
+    firehoseClient.putRecord(
+      {
+        study: 'google-classroom-share-button',
+        study_group: 'v0',
+        event: 'api_load_finished',
+        data_json: JSON.stringify(data)
+      },
+      {includeUserId: true}
+    );
+  };
+}
+
+export function canShowGoogleShareButton(state) {
+  return !!state.googlePlatformApi && state.googlePlatformApi.loaded;
```

Comment:
Question: Why not use `googlePlatformApi.loaded` directly in the connect portion of the react component?

Reply:
we only register the reducer if we meet the section and user requirements for showing the button, so when that's not the case `!!state.googlePlatformApi === false`, and i don't want consumer components to need to know that detail.

also just as a stylistic preference -- the components that consume this don't really care about whether or not some google platform is loaded, they just care about whether they can render the google button.

Reply:
👍 - Thank you!